src/java.base/share/classes/java/util/random/RandomGeneratorFactory.java

```diff
@@ -0,0 +1,524 @@
+ *     System.out.println(rng.nextLong());
+ * }</pre>
+ *
+ * @since   16
+ *
+ * @see java.util.random
+ *
+ */
+public final class RandomGeneratorFactory<T extends RandomGenerator> {
+    /**
+     * Map of provider classes.
+     */
+    private static volatile Map<String, Provider<? extends RandomGenerator>> factoryMap;
+
+    /**
+     * Instance provider class of random number algorithm.
+     */
+    private final Provider<? extends RandomGenerator> provider;
+
+    /**
+     * Map of properties for provider.
+     */
+    private volatile Map<RandomGeneratorProperty, Object> properties = null;
+
+    /**
+     * Default provider constructor.
+     */
+    private volatile Constructor<T> ctor;
+
+    /**
+     * Provider constructor with long seed.
+     */
+    private Constructor<T> ctorLong;
+
+    /**
+     * Provider constructor with byte[] seed.
+     */
+    private Constructor<T> ctorBytes;
+
+    /**
+     * Private constructor.
+     *
+     * @param provider  Provider class to wrap.
+     */
+    private RandomGeneratorFactory(Provider<? extends RandomGenerator> provider) {
+        this.provider = provider;
+    }
+
+    /**
+     * Returns the factory map, lazily constructing map on first call.
+     *
+     * @return Map of RandomGeneratorFactory classes.
+     */
+    private static Map<String, Provider<? extends RandomGenerator>> getFactoryMap() {
+        Map<String, Provider<? extends RandomGenerator>> fm = RandomGeneratorFactory.factoryMap;
+        if (fm == null) {
+            synchronized (RandomGeneratorFactory.class) {
+                if (RandomGeneratorFactory.factoryMap == null) {
+                    fm = RandomGeneratorFactory.factoryMap =
+                        ServiceLoader
+                            .load(RandomGenerator.class)
+                            .stream()
+                            .filter(p -> !p.type().isInterface())
+                            .collect(Collectors.toMap(p -> p.type().getSimpleName().toUpperCase(),
+                                    Function.identity()));
+                }
+            }
+        }
+        return fm;
+    }
+
+    /**
+     * Return the properties map for the specified provider.
+     *
+     * @return properties map for the specified provider.
+     */
+    @SuppressWarnings("unchecked")
+    private Map<RandomGeneratorProperty, Object> getProperties() {
+        Map<RandomGeneratorProperty, Object> props = properties;
+        if (props == null) {
+            synchronized (provider) {
+                if (properties == null) {
+                    try {
+                        Method getProperties = provider.type().getDeclaredMethod("getProperties");
+                        PrivilegedExceptionAction<Map<RandomGeneratorProperty, Object>> getAction = () -> {
+                            getProperties.setAccessible(true);
+                            return (Map<RandomGeneratorProperty, Object>)getProperties.invoke(null);
```

Comment:
Given that we have no control over the fact that the method properties() doesn't return a Map of something else, this unsafe cast seems dangerous (from the type system POV).

Having a type representing a reified version of the Map may be a better idea